Sources/SnapshotTesting/Snapshotting/UIImage.swift

```diff
@@ -169,83 +169,77 @@ private func diff(_ old: UIImage, _ new: UIImage) -> UIImage {
-      inputs: [deltaOutputImage],
-      arguments: [ThresholdImageProcessorKernel.inputThresholdKey: (1 - perceptualPrecision) * 100]
-    )
-  } catch {
-    return "Newly-taken snapshot's data could not be loaded. \(error)"
-  }
-  var averagePixel: Float = 0
+  // Setting the working color space and output color space to NSNull disables color management. This is appropriate when the output
+  // of the operations is computational instead of an image intended to be displayed.
   let context = CIContext(options: [.workingColorSpace: NSNull(), .outputColorSpace: NSNull()])
-  context.render(
-    thresholdOutputImage.applyingFilter("CIAreaAverage", parameters: [kCIInputExtentKey: new.extent]),
-    toBitmap: &averagePixel,
-    rowBytes: MemoryLayout<Float>.size,
-    bounds: CGRect(x: 0, y: 0, width: 1, height: 1),
-    format: .Rf,
-    colorSpace: nil
-  )
-  let actualPixelPrecision = 1 - averagePixel
-  guard actualPixelPrecision < pixelPrecision else { return nil }
+  guard let buffer = deltaOutputImage.render(extent: deltaOutputImage.extent, context: context) else {
+    return "Newly-taken snapshot could not be processed."
+  }
+  defer { buffer.free() }
+  let deltaThreshold = (1 - perceptualPrecision) * 100
+  var failingPixelCount: Int = 0
   var maximumDeltaE: Float = 0
-  context.render(
-    deltaOutputImage.applyingFilter("CIAreaMaximum", parameters: [kCIInputExtentKey: new.extent]),
-    toBitmap: &maximumDeltaE,
-    rowBytes: MemoryLayout<Float>.size,
-    bounds: CGRect(x: 0, y: 0, width: 1, height: 1),
-    format: .Rf,
-    colorSpace: nil
-  )
-  let actualPerceptualPrecision = 1 - maximumDeltaE / 100
-  if pixelPrecision < 1 {
-    return """
-    Actual image precision \(actualPixelPrecision) is less than required \(pixelPrecision)
-    Actual perceptual precision \(actualPerceptualPrecision) is less than required \(perceptualPrecision)
-    """
-  } else {
-    return "Actual perceptual precision \(actualPerceptualPrecision) is less than required \(perceptualPrecision)"
+  // rowBytes must be a multiple of 8, so vImage_Buffer pads the end of each row with bytes to meet the multiple of 0 requirement.
+  // We must do 2D iteration of the vImage_Buffer in order to avoid loading the padding garbage bytes at the end of each row.
+  fastForEach(in: 0..<Int(buffer.height)) { line in
+    let lineOffset = buffer.rowBytes * line
+    fastForEach(in: 0..<Int(buffer.width)) { column in
+      let columnOffset = lineOffset + column * MemoryLayout<Float>.size
+      let deltaE = buffer.data.load(fromByteOffset: columnOffset, as: Float.self)
+      if deltaE > deltaThreshold {
+        failingPixelCount += 1
+      }
+      maximumDeltaE = max(maximumDeltaE, deltaE)
+    }
   }
+  let failingPixelPercent = Float(failingPixelCount) / Float(deltaOutputImage.extent.width * deltaOutputImage.extent.height)
+  let actualPixelPrecision = 1 - failingPixelPercent
+  guard actualPixelPrecision < pixelPrecision else { return nil }
```

Comment:
couldn't this be inside the for loop? we know the size of the `deltaOutputImage` before the loop, so we can calculate the number of pixels that may fail (rather than computing precision every iteration after incrementing failed pixel count) and then the loop can be aborted as soon as this number is exceeded. Otherwise we are comparing all the pixels even when we know that snapshot is not precise enough.
I guess we don't have this problem when snapshots match since they would be compared for exact match before this.

Reply:
I've tried that in my [fork](https://github.com/IlyaPuchkaTW/swift-snapshot-testing/commit/8f873041b3f8af99b71009a41324d8cee2aaa946) but it didn't give me any significant improvements

Reply:
Yeah, as the complexity would not change from O(n) (n = number of pixels) and unless you're comparing a very large images with huge custom frame sizes, the tests times will not improve significantly